[Feature] create `construct_scorecard` and fix `get_leafs` and in `lgb_constructor`

RektPunk opened the pull request:
Related to https://github.com/xRiskLab/xBooster/issues/7

## Changes
- Create `construct_scorecard` method
- Fix `get_leafs` method


## Requesting Feedback
- Logic for Leaf Index Mapping: generating a relative leaf index within each tree (tree_index). This index is computed using groupby('tree_index').cumcount() on the leaf nodes extracted from trees_to_dataframe(), ensuring alignment with the output of predict(pred_leaf=True).

## References
https://stackoverflow.com/questions/78416214/how-many-trees-do-i-actually-have-in-my-lightgbm-model

## Summary by Sourcery

Implement scorecard construction in the LightGBM constructor, fix leaf extraction and indexing logic, and enhance the scorecard with observation percentage metrics

New Features:
- Implement construct_scorecard to generate per-tree scorecards with event counts, rates, WOE, IV, and evidence contributions

Bug Fixes:
- Simplify get_leafs by using predict(start_iteration, num_iteration) and correct base_score subtraction
- Fix extract_leaf_weights to assign a relative leaf index within each tree

Enhancements:
- Add CountPct calculation to report the percentage of observations per split

RektPunk committed:
make get leaf simplify

RektPunk committed:
make constuct scorecard

RektPunk committed:
diff base score for 0 tree

RektPunk committed:
fix typo

RektPunk committed:
make test happy

xRiskLab commented:
## Task: Verify Base Score and Margin Summation

We need to verify that LightGBM's `get_leafs()` correctly returns per-tree margins that sum to the total raw prediction, similar to the verification approach from [Issue #2](https://github.com/xRiskLab/xBooster/issues/2).

**What to verify:**
1. Per-tree margins from `get_leafs()` sum correctly to model's total raw prediction
2. The base_score calculation (`np.log(y.mean() / (1 - y.mean()))`) is appropriate for LightGBM
3. Add a test using a "two-tree trick" to isolate and verify the base margin behavior

**Current status:**
- Fixed `get_leafs()` to return margins without incorrect base_score subtraction
- Example validates that margins sum correctly: `np.allclose(margin_sum, raw_pred)`
- All tests pass

**Recommended before merge:**
- Add explicit test that validates margin summation (similar to two-tree test from Issue #2)
- Document how LightGBM handles base predictions in tree structure
- Verify that `construct_scorecard()` correctly interprets the margin values

xRiskLab committed:
fix: correct get_leafs() margin summation and add base score validation

- Remove incorrect base_score subtraction from get_leafs()
- Add two-tree test to validate margin summation (similar to Issue #2)
- Update documentation explaining LightGBM margin behavior
- Simplify example to show correct margin verification

Fixes margin calculation where tree predictions now correctly sum to total raw score without additional base_score adjustment.

xRiskLab commented:
## ✅ Base Score Verification - COMPLETED

The verification tasks have been completed in commit `8c220f7`:

**What was fixed:**
1. ✅ Removed incorrect `base_score` subtraction from `get_leafs()` 
2. ✅ Added `test_two_tree_base_score_validation()` - the two-tree trick test
3. ✅ Enhanced documentation in `get_leafs()` and `__init__` explaining LightGBM margin behavior
4. ✅ Simplified example to show correct margin verification

**Test results:**
- All 11 LightGBM tests passing
- New two-tree test validates:
  - Per-tree margins sum to total raw prediction ✓
  - Constructor's `get_leafs()` matches LightGBM API ✓
  - Base score calculation matches log-odds ✓

**What to review next:**
@RektPunk - Please review the new commit to ensure the fix aligns with your original implementation. The key change is that LightGBM's per-tree predictions already sum correctly without needing to add/subtract base_score separately (unlike XGBoost).

You can run the tests with: `uv run pytest tests/test_lgb_constructor.py::test_two_tree_base_score_validation -v`

RektPunk commented:
Thank you for creating excellent tests and examples, @xRiskLab. I really appreciate it!!

I am wondering if removing the base score from the first tree is compatible with the `xBooster` logic that was fixed in this commit: https://github.com/xRiskLab/xBooster/commit/8c220f7ed760c403114fd64432a3b86090659556#diff-918f8f30807774e4c31ebfadd85846f60849cb5abe698e612e6be3a0f35755b9L174.

I had thought the base score was added to the first tree, as referenced here: https://github.com/microsoft/LightGBM/issues/3058#issuecomment-668396064.

xRiskLab committed:
docs: clarify sklearn API vs internal booster base_score behavior

- Add note in get_leafs() distinguishing sklearn API from internal booster
- Reference LightGBM issue #3058 for internal booster behavior
- Add lightgbm-getting-started.ipynb demonstrating the difference empirically

RektPunk commented:
Hi @xRiskLab , Could you please confirm if there are any further changes needed? If not, is it ready to merge now?

xRiskLab commented:
## ⚠️ Update: Release v0.2.7rc2 (Complete Implementation)

**Issue Found**: The squash merge only included the original PR commits, missing the complete implementation.

**Resolution**: Released **v0.2.7rc2** with the complete implementation:
- ✅ `create_points()` with `use_base_score` parameter  
- ✅ `predict_score()` and `predict_scores()`
- ✅ Critical leaf ID mapping fix
- ✅ All 106 tests passing

### 📦 Installation
```bash
pip install xbooster==0.2.7rc2
```

### 🔗 Links
- **GitHub Release**: https://github.com/xRiskLab/xBooster/releases/tag/v0.2.7rc2
- **PyPI**: https://pypi.org/project/xbooster/0.2.7rc2/
